### PR TITLE
DEV: add function used at srx for getting data from db

### PR DESCRIPTION
**Pull request by licode**
This is one of the backend dataIO functions used at SRX. This function deals with step scan, fly scan and also the cases when scan not finished. The overall structure looks not clean, but that's the complexity we face for now.

Test will also be added.

### CODE REVIEW

File: suitcase/pyxrf/export.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    """
+    spectrum_len = 4096
+    start_doc = hdr['start']
+    plan_n = start_doc.get('plan_name')
+    if 'fly' not in plan_n: # not fly scan, step scan instead
+        datashape = start_doc['shape']   # vertical first then horizontal
+        fly_type = None
+
+        snake_scan = start_doc.get('snaking')
+        if snake_scan[1] == True:
```
Comment (danielballan) [inline suitcase/pyxrf/export.py]: `if snake_scan[1] == True` should be just `if snake_scan[1]`

File: suitcase/pyxrf/export.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            fly_type = 'pyramid'
+
+        try:
+            data = hdr.table(fill=True, convert_times=False)
+        except IndexError: # scan is not finished
```
Comment (danielballan) [inline suitcase/pyxrf/export.py]: Could we check for the existence of a stop document instead, using `hdr.stop is not None`? Catching this `IndexError` could mask problems other than "scan is not finished".

File: suitcase/pyxrf/export.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+            total_len = get_total_scan_point(hdr) - 2
+            evs, _ = zip(*zip(hdr.events(hdr, fill=True), range(total_len)))
+            namelist = (config_data['xrf_detector'] +
```
Comment (danielballan) [inline suitcase/pyxrf/export.py]: Is this related to `hdr.config_data` or just a coincidence that it has the same name?

File: suitcase/pyxrf/export.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                        hdr.start.motors +config_data['scaler_list'])
+            dictv = {v:[] for v in namelist}
+            for e in evs:
+                for k,v in six.iteritems(dictv):
```
Comment (danielballan) [inline suitcase/pyxrf/export.py]: No need for Python 2 support; you can just do `dictv.items()`.

File: suitcase/pyxrf/export.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+                                  fly_type=fly_type, spectrum_len=spectrum_len)
+            write_db_to_hdf_base(fpath1, new_data, num_det=num_det,
+                                 create_each_det=create_each_det)
+    else:
+        # srx fly scan
+        # Added by AMK to allow flying of single element on xs2
+        if 'E_tomo' in start_doc['scaninfo']['type']:
+            num_det = 1
+
+        if save_scalar is True:
```
Comment (danielballan) [inline suitcase/pyxrf/export.py]: The Python way to write this is `if save_scalar` so that any "truth-y" value will do. There are several cases like this below with the same issue.

File: suitcase/pyxrf/export.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+            dataGrp = f.create_group(interpath+'/scalers')
+            scaler_names = data['scaler_names']
+            scaler_data = data['scaler_data']
+            dataGrp.create_dataset('name', data=helper_encode_list(scaler_names))
+            dataGrp.create_dataset('val', data=scaler_data)
+
+
+def get_name_value_from_db(name_list, data, datashape):
+    """
+    Get data from db, and parse it to the required format and output.
```
Comment (danielballan) [inline suitcase/pyxrf/export.py]: I would rephrase the docstring and maybe rename the function. "Get data from the db" sounds like it fetch data from the DataBroker.

File: suitcase/pyxrf/export.py [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+    scaler_names, scaler_data = get_name_value_from_db(scaler_list, data,
+                                                       datashape)
+    if fly_type in ('pyramid',):
+        scaler_data = flip_data(scaler_data, subscan_dims=subscan_dims)
+    for i, v in enumerate(scaler_names):
+        data_output[v] = scaler_data[:, :, i]
+    return data_output
+
+
+def write_db_to_hdf_base(fpath, data, num_det=3, create_each_det=True):
```
Comment (danielballan) [inline suitcase/pyxrf/export.py]: What does `_base` mean?

**Review by danielballan — COMMENTED** [review]
Left some code-style comments. Still digesting the big picture.

File: suitcase/pyxrf/export.py (hunk 2 above)
Comment (licode) [reply suitcase/pyxrf/export.py]: agree. this is mainly to catch hdr.stop issue.

File: suitcase/pyxrf/export.py (hunk 7 above)
Comment (licode) [reply suitcase/pyxrf/export.py]: no need to keep base here. somewhat confusing.

File: suitcase/pyxrf/export.py (hunk 3 above)
Comment (licode) [reply suitcase/pyxrf/export.py]: maybe coincidence. The goal is to add those information to header later through RE. Now I just pass config_data dictionary as input to this function.

File: suitcase/pyxrf/export.py (hunk 6 above)
Comment (licode) [reply suitcase/pyxrf/export.py]: I will add more doc. thanks for suggestion.

File: .travis.yml [hunk 8]
```diff
@@ -9,6 +8,7 @@ cache:
 
 install:
   - pip install .
+  - pip install -r requirements.txt
```
Comment (danielballan) [inline .travis.yml]: This should not be necessary. The `pip install .` above runs the `setup.py` which reads and parses `requirements.txt` and passes the entries to `setup(..., install_requires=requirements)`. Make sense?

File: run_tests.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
+#!/usr/bin/env python
```
Comment (danielballan) [inline run_tests.py]: Do we need this? I've come to think our projects don't need a special `run_tests.py`, that running `pytest` directly is simpler, which is why I didn't include it in the cookiecutters. Any configuration/defaults we want to include could be put in the pytest configuration file.

File: .travis.yml [hunk 10]
```diff
@@ -1,6 +1,5 @@
 language: python
 python:
-  - 3.5
```
Comment (danielballan) [inline .travis.yml:3]: I think we should support 3.5 for now, since databroker does. Any particular reason not to? We should be consistent about this, anyway.

**Review by danielballan — COMMENTED** [review]
Looking cleaner! In addition to my comments/questions posted in line, I noticed that you have committed a binary file named `suitcase/tests/__pycache__/test_srx_export.cpython-36-PYTEST.pyc`, presumably by mistake.

File: .travis.yml (hunk 8 above)
Comment (licode) [reply .travis.yml]: 👍

File: .travis.yml (hunk 10 above)
Comment (licode) [reply .travis.yml:3]: sure. I am fine either way. I just found that we drop 3.5 for nsls2-tag. I will put it back here.

Comment (danielballan) [reply .travis.yml:3]: I see. Yes, we don't build packages for it because we don't deploy 3.5 at NSLS-II, but our packages still support 3.5. I believe some of our collaborators still rely on that.

File: run_tests.py (hunk 9 above)
Comment (licode) [reply run_tests.py]: I probably copied this from other repos. We use pytest now, so I can help to clean up run_test.py in other repos too.

File: suitcase/pyxrf/export.py [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+    for i, v in enumerate(name_list):
+        # keep shape unchanged, so stopped/aborted run can be handled.
+        posv = np.zeros(datashape[0]*datashape[1])
+        data[v] = np.asarray(data[v])
+        posv[:data[v].shape[0]] = np.asarray(data[v])  # scan may not finish.
+        pos_data[:, :, i] = posv.reshape([datashape[0], datashape[1]])
+    return pos_data
+
+
+def flip_data(input_data, subscan_dims=None):
```
Comment (danielballan) [inline suitcase/pyxrf/export.py:378]: My initial reaction to this is that analysis/processing code should stay out of this I/O library. Is there another place we can put this?

File: requirements.txt [hunk 12]
```diff
@@ -0,0 +1,2 @@
+h5py
+pandas
```
Comment (danielballan) [inline requirements.txt:2]: I can't find `import pandas` in here. Is it actually used?

**Review by danielballan — COMMENTED** [review]
I took another quick look. I think another stray file was accidentally committed: `suitcase/tests/__pycache__/test_srx_export.cpython-36-PYTEST.pyc`

**Comment by licode** [comment]
Can I close this PR, and send less functions with test in a new PR? So it is easy to review?

**Comment by danielballan** [comment]
Sounds good.
